workflow/snakemake_rules/fetch_sequences.smk

```diff
@@ -29,48 +30,72 @@ def run_shell_command_n_times(cmd, msg, cleanup_failed_cmd, retry_num=5):
     output:
         ndjson = temp(f"data/{database}.ndjson")
     run:
-        if config.get("fetch_from_database", False):
-            if database=="gisaid":
-                msg = "Fetching from GISAID API"
-                cmd = f"./bin/fetch-from-gisaid {output.ndjson}"
-            else:
-                msg = "Fetching from GenBank API"
-                cmd = f"./bin/fetch-from-genbank > {output.ndjson}"
-            cleanup_failed_cmd = f"rm {output.ndjson}"
-            run_shell_command_n_times(cmd, msg, cleanup_failed_cmd)
-            if send_notifications:
-                shell("./bin/notify-on-record-change {output.ndjson} {params.file_on_s3_src} {database}")
+        if database == "gisaid":
+            cmd = f"./bin/fetch-from-gisaid {output.ndjson}"
         else:
-            shell("""
-                ./bin/download-from-s3 {params.file_on_s3_dst} {output.ndjson} ||  \
-                ./bin/download-from-s3 {params.file_on_s3_src} {output.ndjson}
-            """)
+            cmd = f"./bin/fetch-from-genbank > {output.ndjson}"
 
-rule download_biosample:
+        run_shell_command_n_times(
+            cmd,
+            f"Fetching from {database}",
+            f"rm {output.ndjson}"
+        )
+
+rule fetch_biosample:
     message:
-        """Obtaining Biosample data (GenBank only)"""
-    params:
-        file_on_s3_dst = config["s3_dst"] + '/biosample.ndjson.gz',
-        file_on_s3_src = config["s3_src"] + '/biosample.ndjson.gz',
+        """Fetching BioSample data (GenBank only)"""
     output:
-        biosample = "data/biosample.ndjson"
+        biosample = temp("data/biosample.ndjson")
     run:
-        if config.get("fetch_from_database", False):
-            run_shell_command_n_times(
-                f"./bin/fetch-from-biosample > {output.biosample}",
-                "Fetch BioSample",
-                f"rm {output.biosample}")
-        else:
-            shell("""
-                ./bin/download-from-s3 {params.file_on_s3_dst} {output.biosample} ||  \
-                ./bin/download-from-s3 {params.file_on_s3_src} {output.biosample}
-            """)
+        run_shell_command_n_times(
+            f"./bin/fetch-from-biosample > {output.biosample}",
+            "Fetch BioSample",
+            f"rm {output.biosample}"
+        )
+
+# Only include rules to fetch from S3 if S3 config params are provided
+if config.get("s3_dst") and config.get("s3_src"):
+
+    # Set ruleorder since these fetch rules have the same output
+    # Fetch directly from databases when `fetch_from_database=True`
+    # or else fetch files from AWS S3 buckets
+    if config.get("fetch_from_database", False):
+        ruleorder: fetch_main_ndjson > fetch_main_ndjson_from_s3
+        ruleorder: fetch_biosample > fetch_biosample_from_s3
+    else:
+        ruleorder: fetch_main_ndjson_from_s3 > fetch_main_ndjson
+        ruleorder:  fetch_biosample_from_s3 > fetch_biosample
```

Comment:
_non-blocking_

An alternative to defining two sets of rules and then using `ruleorder` to disambiguate them is to conditionally define the rules, e.g.:

```python
if config.get("fetch_from_database", False):
    rule fetch_main_ndjson:
        # fetch from upstream database
    rule fetch_biosample:
        # fetch from upstream database

elif config.get("s3_dst") and config.get("s3_src"):
    rule fetch_main_ndjson:
        # fetch from S3
    rule fetch_biosample:
        # fetch from S3
```

This might make things easier? It does mean you can't, for example, directly run `rule fetch_main_ndjson_from_s3` even when `fetch_from_database` is true, but I don't think we need to do that.

This also brings to mind that the conditional used by the replaced rules:

```
run:
    if config.get("fetch_from_database", False):
        # fetch from upstream database
    else:
        # fetch from S3
```

could have alternatively to this change been modified from `else` → `elif config.get("s3_dst") and config.get("s3_src")`.

Is there a reason the separate rules are preferable? They seem cleaner, but not sure about other reasons.

Reply:
I can see how conditionally defined rules is cleaner. I wanted to make the fetch from upstream databases the default behavior even when nothing is specified in the config, which is why I went with the current approach.

Reply:
That makes sense.